File: apps/mobile/android/src/main/kotlin/com/messenger/store/BiometricLockStore.kt
package com.messenger.store

import android.content.Context
import android.content.SharedPreferences
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import java.security.MessageDigest

data class AppLockSettings(
    val enabled: Boolean = true,
    val relockTimeoutSeconds: Int = 0, // 0 = немедленно; -1 = никогда
    val pinHashSha256: String? = null,
)

object BiometricLockStore {
    private lateinit var prefs: SharedPreferences

    private val _settings = MutableStateFlow(AppLockSettings())
    val settings: StateFlow<AppLockSettings> = _settings

    private val _isLocked = MutableStateFlow(true)
    val isLocked: StateFlow<Boolean> = _isLocked

    fun init(context: Context) {
        // Используем обычные SharedPreferences (не EncryptedSharedPreferences чтобы не добавлять зависимость)
        prefs = context.getSharedPreferences("biometric_lock", Context.MODE_PRIVATE)
        _settings.value = AppLockSettings(
            enabled = prefs.getBoolean("enabled", true),
            relockTimeoutSeconds = prefs.getInt("relock_timeout", 0),
            pinHashSha256 = prefs.getString("pin_hash", null),
        )
    }

    fun unlock() { _isLocked.value = false }
    fun lock() { _isLocked.value = true }

    fun isPinCorrect(pin: String): Boolean {
        val hash = _settings.value.pinHashSha256 ?: return pin.length >= 4
        return sha256(pin) == hash
    }

    fun saveSettings(settings: AppLockSettings) {
        _settings.value = settings
        prefs.edit()
            .putBoolean("enabled", settings.enabled)
            .putInt("relock_timeout", settings.relockTimeoutSeconds)
            .apply()
    }

    fun updatePin(newPin: String) {
        val hash = sha256(newPin)
        prefs.edit().putString("pin_hash", hash).apply()
        _settings.value = _settings.value.copy(pinHashSha256 = hash)
    }

    private fun sha256(input: String): String {
        val bytes = MessageDigest.getInstance("SHA-256").digest(input.toByteArray())
        return bytes.joinToString("") { "%02x".format(it) }
    }
}

File: apps/mobile/ios/Sources/Messenger/store/BiometricLockStore.swift
import Foundation
import CryptoKit

struct AppLockSettings: Codable {
    var enabled: Bool = true
    var relockTimeoutSeconds: Int = 0
    var pinHashSha256: String? = nil
}

final class BiometricLockStore: ObservableObject {
    static let shared = BiometricLockStore()
    private static let key = "messenger.lock.settings"

    @Published var settings: AppLockSettings
    @Published var isLocked: Bool = true

    private init() {
        if let data = UserDefaults.standard.data(forKey: Self.key),
           let decoded = try? JSONDecoder().decode(AppLockSettings.self, from: data) {
            settings = decoded
        } else {
            settings = AppLockSettings()
        }
    }

    func unlock() { isLocked = false }
    func lock() { isLocked = true }

    func isPinCorrect(_ pin: String) -> Bool {
        guard let hash = settings.pinHashSha256 else { return pin.count >= 4 }
        return sha256(pin) == hash
    }

    func saveSettings(_ newSettings: AppLockSettings) {
        settings = newSettings
        if let data = try? JSONEncoder().encode(newSettings) {
            UserDefaults.standard.set(data, forKey: Self.key)
        }
    }

    func updatePin(_ newPin: String) {
        let hash = sha256(newPin)
        settings.pinHashSha256 = hash
        saveSettings(settings)
    }

    private func sha256(_ input: String) -> String {
        let data = Data(input.utf8)
        let digest = SHA256.hash(data: data)
        return digest.map { String(format: "%02x", $0) }.joined()
    }
}
